### PR TITLE
fix(recipes): Allow `undefined` for variants

### LINKED ISSUE #1442: Recipe's runtime functions do not accept `undefined` as valid values for variants at the type level (`exactOptionalPropertyTypes`)
### Describe the bug

With TypeScript's `exactOptionalPropertyTypes` enabled, recipe's runtime functions do not accept `undefined` as valid values for variants at the type level.

### Reproduction

- Add `"exactOptionalPropertyTypes": true` to `compilerOptions` in tsconfig.json.
- Copy the code below into a typescript file.
- Observe the error in your editor or run `pnpm lint:tsc`.

```ts
import { recipe } from "@vanilla-extract/recipes";

const r = recipe({
	variants: {
		color: {
			neutral: { background: "whitesmoke" },
		},
	},
});

const color: "neutral" | undefined = undefined as "neutral" | undefined;

r({ color });
```

_Without_ `exactOptionalPropertyTypes` `r`'s type is
`(options?: { color?: "neutral" | undefined; } | undefined) => string`.
_With_ it enabled the type changes to
`(options?: { color?: "neutral"; } | undefined) => string`,
therefore the example above doesn't satisfy the type checker.

```
tests/recipes/test.ts:13:3 - error TS2379: Argument of type '{ color: "neutral" | undefined; }' is not assignable to parameter of type '{ color?: "neutral"; }' with 'exactOptionalPropertyTypes: true'. Consider adding 'undefined' to the types of the target's properties.
  Types of property 'color' are incompatible.
    Type '"neutral" | undefined' is not assignable to type '"neutral"'.
      Type 'undefined' is not assignable to type '"neutral"'.

r({ color });
```

### System Info

I'm using typescript 5.5.3, though the problem should affect all typescript versions since 4.4.0.

### Validations

- [X] Check that there isn't [already an issue](https://github.com/vanilla-extract-css/vanilla-extract/issues) that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

### PR DESCRIPTION
**Pull request by DominikRusso**
Closes #1442

### CODE REVIEW

**Comment by DominikRusso** [comment]
Hi @askoufis, is there anything I can do to get this reviewed and merged any quicker?

File: .changeset/tame-suits-refuse.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+'@vanilla-extract/recipes': patch
+---
+
+Allow passing `undefined` for variants to runtime functions
+when TypeScript's `exactOptionalPropertyTypes` is enabled.
```
Comment (askoufis) [inline .changeset/tame-suits-refuse.md:6]:
```suggestion
Enable passing `undefined` variant values to recipe functions when TypeScript's `exactOptionalPropertyTypes` is enabled.
```

**Review by askoufis — APPROVED** [review]
Sorry for the delay. Just a minor comment on the changeset wording.

**Comment by DominikRusso** [comment]
@askoufis thank you, i fixed up the changes

**Comment by askoufis** [comment]
> @askoufis thank you, i fixed up the changes

Thanks. Your branch is also out-of-date relative to `master`. Ordinarily I'd be able to update it, but maybe you have some stricter permissions on your fork?

**Comment by DominikRusso** [comment]
@askoufis ah sorry, i think it's because i had removed the *Allow edits by maintainers* checkmark. should be rebased now

**Pull request merged**